core/src/main/java/org/infinispan/stream/impl/IteratorHandler.java

```diff
@@ -0,0 +1,244 @@
+      return currentRequests.size();
+   }
+
+   /**
+    * Invoked to have handler forget about given iterator they requested. This should be invoked when an
+    * {@link Iterator} has been found to have completed {@link Iterator#hasNext()} is false or if the caller
+    * wishes to stop a given iteration early. If an iterator is fully iterated upon (ie. {@link Iterator#hasNext()}
+    * returns false, that invocation will also close resources related to the iterator.
+    * @param requestId the id of the iterator
+    */
+   public void closeIterator(Address origin, Object requestId) {
+      Set<Object> ids = ownerRequests.get(origin);
+      if (ids != null) {
+         ids.remove(requestId);
+      }
+      closeIterator(requestId);
+   }
+
+   private void closeIterator(Object requestId) {
+      CloseableIterator<?> closeableIterator = currentRequests.remove(requestId);
+      if (closeableIterator != null) {
+         if (trace) {
+            log.tracef("Closing iterator using requestId %s", requestId);
+         }
+         closeableIterator.close();
+      }
+   }
+
+   private class IteratorCloser<E> implements OnCloseIterator<E> {
+      private final CloseableIterator<E> closeableIterator;
+      private volatile Runnable closeRunnable;
+
+      IteratorCloser(CloseableIterator<E> closeableIterator) {
+         this.closeableIterator = closeableIterator;
+      }
+
+      @Override
+      public boolean hasNext() {
+         boolean hasNext = closeableIterator.hasNext();
+         if (!hasNext) {
+            close();
+         }
+         return hasNext;
+      }
+
+      @Override
+      public E next() {
+         // Make sure to double check hasNext in case if user didn't call it before next.
+         hasNext();
+         return closeableIterator.next();
+      }
+
+      @Override
+      public void forEachRemaining(Consumer<? super E> action) {
+         closeableIterator.forEachRemaining(action);
+         close();
+      }
+
+      @Override
+      public void close() {
+         closeableIterator.close();
+         // We only want to run the runnable once, in case if we end up closing twice
+         Runnable onClose = closeRunnable;
+         if (onClose != null) {
+            closeRunnable = null;
+            onClose.run();
+         }
+      }
+
+      @Override
+      public IteratorCloser<E> onClose(Runnable closeHandler) {
+         if (this.closeRunnable == null) {
+            this.closeRunnable = closeHandler;
+         } else {
+            this.closeRunnable = composeWithExceptions(this.closeRunnable, closeHandler);
+         }
+         return this;
+      }
+   }
+
+   /**
+    * Given two Runnables, return a Runnable that executes both in sequence,
+    * even if the first throws an exception, and if both throw exceptions, add
+    * any exceptions thrown by the second as suppressed exceptions of the first.
+    */
+   private static Runnable composeWithExceptions(Runnable a, Runnable b) {
```

Comment:
Is it worth creating a StreamUtil class for code like this that is repeated?

Reply:
It isn't really specific to streams per se. I can add it to `commons.util.Util` as this is really just a Java 8 thing that allows composition of exceptions when combining runnables.

Reply:
Sounds good